server/tests/fstar_handlers.test.ts

```diff
@@ -0,0 +1,124 @@
+	});
+
+	test('test fragmented message', () => {
+		const fragment_0 = '{"kind": "';
+		const fragment_1 = 'message';
+		const fragment_2 = '"}';
+
+		// Fragments are buffered until a full valid message is received, then it is
+		// passed on for further handling.
+		handleFStarResponseForDocument(td, fragment_0, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(0);
+		handleFStarResponseForDocument(td, fragment_1, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(0);
+		handleFStarResponseForDocument(td, fragment_2, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(1);
+		expect(handleOneResponseForDocumentMock).toHaveBeenLastCalledWith(td, fragment_0 + fragment_1 + fragment_2, lax);
+	});
+
+	test('test out-of-order fragmented messages are not handled', () => {
+		const fragment_0 = '{"kind": "';
+		const fragment_1 = 'message';
+		const fragment_2 = '"}';
+
+		// Fragments are assumed to be received in-order, so out-of-order
+		// fragments have undefined behavior. In this test case, no valid
+		// message can be collected.
+		handleFStarResponseForDocument(td, fragment_2, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(0);
+		handleFStarResponseForDocument(td, fragment_1, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(0);
+		handleFStarResponseForDocument(td, fragment_0, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(0);
+	});
+
+	test('test valid messages flush buffer', () => {
+		const valid_message = '{"kind": "message"}';
+		const fragment_0 = '{"kind": "';
+		const fragment_1 = 'message';
+		const fragment_2 = '"}';
+
+		// Fragments are assumed to be received in-order and before other
+		// messages, so a valid message results in the buffer being flushed.
+		handleFStarResponseForDocument(td, valid_message, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(1);
+		expect(handleOneResponseForDocumentMock).toHaveBeenLastCalledWith(td, valid_message, lax);
+
+		handleFStarResponseForDocument(td, fragment_0, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(1);
+
+		handleFStarResponseForDocument(td, valid_message, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(2);
+		expect(handleOneResponseForDocumentMock).toHaveBeenLastCalledWith(td, valid_message, lax);
+
+		handleFStarResponseForDocument(td, fragment_1, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(2);
+		handleFStarResponseForDocument(td, fragment_2, lax);
+		expect(handleOneResponseForDocumentMock).toHaveBeenCalledTimes(2);
+
+	});
+
+	test('test combined messages and fragments processed separately', () => {
+		const valid_message = '{"kind": "message"}';
+		const fragment_0 = '{"kind": "';
+		const fragment_1 = 'message';
+		const fragment_2 = '"}';
+
+		const combined_messages = [valid_message, fragment_0, fragment_1, fragment_2].join('\n');
```

Comment:
Good point, this case should never happen. My preference is to retain this test since it helps document the current fragmentation handling behavior I implemented (and would be safe in case the F* IDE code changed behavior), but the test could be removed. Would you prefer to see the test removed?

Reply:
See my comment below.  I think we can keep the test, but it should test behavior that can actually happen.

Reply:
At this point I think I should do the refactoring in a separate PR that can be merged quickly. I opened https://github.com/FStarLang/fstar-vscode-assistant/issues/25 to track this issue.